### PR TITLE
[Transformers future] Loss Computation for Compatibility with Transformers 4.48.3

### PR DESCRIPTION
**Pull request by yafshar**
# What does this PR do?

- Update the GaudiTrainer to better match with with Transformers 4.48.3 with minor enhancement from 4.49.0
  - Remove the `get_batch_samples_transformers`
  - Implemented `get_num_items_in_batches` to calculate the number of items in each batch for all epochs during training.
  - Added `get_iterator_batch_samples` to retrieve batch samples from an iterator. 
  - Moved the calculation of `num_items_in_batches` outside the training loop to recover performance lost. This feature should be up streamed to Transformers branch.
 
- Update llama model loss computation
  - Simplified loss computation by using a configurable loss function.
  - Replaced the inline loss computation logic with a call to self.loss_function.
  - Introduced ForCausalLMContextParallelLoss for context parallel loss computation.
  - Updated the __init__ method to set the loss function based on the parallel strategy.
  - Ensured compatibility with Transformers 4.48.3 by aligning with its structure and conventions.

- Refactor loss computation in context parallel to better match with the upstream transformers
  - Replaced _ContextParallelLoss with ContextParallelLossFunction for better clarity and consistency.
  - Updated fixed_cross_entropy to use ContextParallelLossFunction for  gathering losses across context parallel groups.
  - Introduced ForCausalLMContextParallelLoss to handle loss computation  for causal language modeling with context parallelism.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by yafshar** [comment]
@regisss I removed the  `get_batch_samples_transformers` in GaudiTrainer and use the super class function `get_batch_samples`. The `num_items_in_batch` is only constant for special cases of using examples with `--dataset_concatenation`. That can be an extra optimization for later, but it is not a general case.

File: optimum/habana/transformers/trainer.py [hunk 1]
```diff
@@ -450,6 +450,9 @@ def _tune_save_checkpoint(self, checkpoint_dir: str):
         output_dir = os.path.join(checkpoint_dir, f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}")
         self.save_model(output_dir, _internal_call=True)
         if self.args.should_save:
+            # TODO
+            # Update the `TrainerControl` state to where we are currently
+            # self.state.stateful_callbacks["TrainerControl"] = self.control.state()
```
Comment (regisss) [inline optimum/habana/transformers/trainer.py:455]: I had missed that change! Did you comment it because it doesn't work?

File: optimum/habana/transformers/trainer.py [hunk 2]
```diff
@@ -992,10 +1003,12 @@ def hpu_deepspeed_checkpointing(function, *checkpoint_args, use_reentrant: Optio
                 remainder = args.gradient_accumulation_steps
             update_step = -1
             total_updates = steps_in_epoch // args.gradient_accumulation_steps + 1
+            if args.gradient_accumulation_steps == 1:
+                total_updates -= 1
             for _ in range(total_updates):
                 update_step += 1
                 num_batches = args.gradient_accumulation_steps if update_step != (total_updates - 1) else remainder
-                batch_samples, num_items_in_batch = self.get_batch_samples_transformers(epoch_iterator, num_batches)
+                batch_samples, num_items_in_batch = self.get_batch_samples(epoch_iterator, num_batches)
```
Comment (regisss) [inline optimum/habana/transformers/trainer.py:1011]: This will probably lead to an error in TRL examples as it is pinned to a quite old version: https://github.com/huggingface/optimum-habana/blob/main/examples/trl/requirements.txt#L1
I'll take a look to see if we can update it.

**Review by regisss — COMMENTED** [review]
I left a few comments. Can you also add in the description of the PR that it brings some changes from Transformers 4.49.0 too?

File: optimum/habana/transformers/trainer.py (hunk 1 above)
Comment (yafshar) [reply optimum/habana/transformers/trainer.py:455]: I added a TODO comment because it works with the new accelerator and we are on 0.33.0. However, I haven't tested or integrated it yet. We should definitely transition to using it.

File: optimum/habana/transformers/trainer.py [hunk 3]
```diff
@@ -1351,7 +1365,7 @@ def _maybe_log_save_evaluate(self, tr_loss, _grad_norm, model, trial, epoch, ign
             self._globalstep_last_logged = self.state.global_step
             self.store_flos()
 
-            self.log(logs, start_time=start_time)
```
[the remark this answers is not in the document]
Comment (yafshar) [reply optimum/habana/transformers/trainer.py:1354]: I did not look at TRL examples, will check more

File: optimum/habana/transformers/trainer.py (hunk 1 above)
Comment (regisss) [reply optimum/habana/transformers/trainer.py:455]: Sounds good :+1: 
We can wait for native Gaudi support in Accelerate which should be merged soon: https://github.com/huggingface/accelerate/pull/3378/

File: optimum/habana/transformers/trainer.py [hunk 4]
```diff
@@ -2615,31 +2636,3 @@ def _zero_model_grad(self, model):
             except TypeError:
                 model.zero_grad()
                 model._zero_grad_kwargs = {}
-
-    def get_batch_samples_transformers(self, epoch_iterator, num_batches):
```
[the remark this answers is not in the document]
Comment (regisss) [reply optimum/habana/transformers/trainer.py:2619]: You can try:
```
RUN_SLOW=1 GAUDI2_CI=1 pytest tests/test_examples.py -v -s -k "test_run_clm_gpt2_single_card"
```
Note that you'll have to uncomment this line: https://github.com/yafshar/optimum-habana/blob/transformers_future/tests/utils.py#L38
Using GPT2 because the test is much faster than other tests with bigger models.

File: optimum/habana/transformers/trainer.py (hunk 2 above)
Comment (regisss) [reply optimum/habana/transformers/trainer.py:1011]: We would need TRL v0.15 to not have a clash between the two `self.get_batch_samples` methods of the trainers, but it requires Accelerate >= 0.34.
My suggestion is to keep `get_batch_samples_transformers` and we can add a TODO saying that this should be removed when the Accelerate dependency is upgraded (which should happen soon, see my comments [above](https://github.com/huggingface/optimum-habana/pull/1794#discussion_r1967806946)).

Comment (yafshar) [inline optimum/habana/transformers/trainer.py:1011]: I haven't been able to reproduce the issue. However, I noticed that others have reported similar problems https://github.com/huggingface/trl/issues/2275

what about changing that 
```py
               batch_samples, num_items_in_batch = super(Trainer, self).get_batch_samples(epoch_iterator, num_batches)
```

File: optimum/habana/transformers/trainer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yafshar) [reply optimum/habana/transformers/trainer.py:2619]: The regression for this model and test is > 50% difference. I am looking into it. Thanks for the hint

File: optimum/habana/transformers/trainer.py (hunk 2 above)
Comment (regisss) [reply optimum/habana/transformers/trainer.py:1011]: That would be a better way of managing it for sure! I guess it will ultimately depend on whether or not we need to override `get_batch_samples` to avoid throughput regressions (linked to the [comment above](https://github.com/huggingface/optimum-habana/pull/1794#discussion_r1968517161)).

**Comment by yafshar** [comment]
@regisss with the latest update commit [b3a29ed](https://github.com/huggingface/optimum-habana/pull/1794/commits/b3a29ed9af7ced96a9f2a664ba7d196781e17a89) I recovered the lost performance. 

I will do more tests, but I think this change should also be up streamed to transformers main branch. There is still room for improvement.

File: optimum/habana/transformers/trainer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yafshar) [reply optimum/habana/transformers/trainer.py:2619]: Latest update resolved the performance regression!

**Comment by yafshar** [comment]
```sh
>>> python3 /root/optimum-habana/examples/language-modeling/run_clm.py --model_name_or_path gpt2 --gaudi_config_name Habana/gpt2 --dataset_name wikitext --do_train --output_dir /tmp/tmp66y1mz5g --overwrite_output_dir --learning_rate 0.0002 --per_device_train_batch_size 16 --per_device_eval_batch_size 4 --num_train_epochs 2 --use_habana --throughput_warmup_steps 3 --save_strategy no --use_lazy_mode --do_eval --dataset_config_name wikitext-2-raw-v1 --use_hpu_graphs_for_inference
```

[transformers_future](https://github.com/huggingface/optimum-habana/tree/transformers_future)

```sh
***** train metrics *****
  epoch                       =        2.0
  max_memory_allocated (GB)   =      89.31
  memory_allocated (GB)       =      25.93
  total_flos                  =  2258261GF
  total_memory_available (GB) =      94.62
  train_loss                  =     3.1423
  train_runtime               = 0:00:50.94
  train_samples               =       2318
  train_samples_per_second    =    121.727
  train_steps_per_second      =      7.615
***** eval metrics *****
  epoch                           =        2.0
  eval_accuracy                   =     0.4193
  eval_graph_compliation_duration =     1.6266
  eval_loss                       =     3.0506
  eval_runtime                    = 0:00:03.65
  eval_samples                    =        240
  eval_samples_per_second         =    112.608
  eval_steps_per_second           =     28.152
  max_memory_allocated (GB)       =      89.31
  memory_allocated (GB)           =      25.67
  perplexity                      =    21.1271
  total_memory_available (GB)     =      94.62
```

This PR after the last commit https://github.com/huggingface/optimum-habana/pull/1794/commits/b3a29ed9af7ced96a9f2a664ba7d196781e17a89

```sh
***** train metrics *****
  epoch                       =        2.0
  max_memory_allocated (GB)   =      89.37
  memory_allocated (GB)       =      25.93
  total_flos                  =  2258261GF
  total_memory_available (GB) =      94.62
  train_loss                  =     3.1352
  train_runtime               = 0:00:57.10
  train_samples               =       2318
  train_samples_per_second    =    121.732
  train_steps_per_second      =      7.615
***** eval metrics *****
  epoch                           =        2.0
  eval_accuracy                   =     0.4198
  eval_graph_compliation_duration =     1.5897
  eval_loss                       =     3.0437
  eval_runtime                    = 0:00:03.69
  eval_samples                    =        240
  eval_samples_per_second         =     108.53
  eval_steps_per_second           =     27.132
  max_memory_allocated (GB)       =      89.37
  memory_allocated (GB)           =      25.68
  perplexity                      =    20.9826
  total_memory_available (GB)     =      94.62
```

File: optimum/habana/transformers/trainer.py [hunk 5]
```diff
@@ -1623,6 +1652,7 @@ def training_step(
             `torch.Tensor`: The tensor with training loss on this batch.
         """
         model.train()
+        # TODO
```
[the remark this answers is not in the document]
Comment (yafshar) [reply optimum/habana/transformers/trainer.py:1655]: I have not checked the `self.optimizer.train()` which has been added recently, so the TODO is just a matter of testing, nothing more

[the remark this answers is not in the document]
Comment (yafshar) [reply optimum/habana/transformers/trainer.py:1655]: I was stopped by some performance issue due the latest upgrade. Now the performance is recovered, I can test this and remove the TODO

[the remark this answers is not in the document]
Comment (Chris-Sigopt) [reply optimum/habana/transformers/trainer.py:1655]: Is there a ticket or some work associated with it that this TODO could mention? I just hate it when blank TODO's are in my code, because I can never remember which was waiting on what, if this is going to be a while. 

Or is this todo going to be resolved before transformers_future hits main?

[the remark this answers is not in the document]
Comment (yafshar) [reply optimum/habana/transformers/trainer.py:1655]: No there is no ticket. It is just a new feature and have not been tested. I agree with you I also hate it with blank TODO in the code. I will update this. I was waiting on @regisss to see if he has any comment, otherwise, I will update the rest and remove the TODO or open a ticket if I encounter an issue

**Review by regisss — APPROVED** [review]
Nice!

**Comment by regisss** [comment]
> @regisss with the latest update commit [b3a29ed](https://github.com/huggingface/optimum-habana/pull/1794/commits/b3a29ed9af7ced96a9f2a664ba7d196781e17a89) I recovered the lost performance.
> 
> I will do more tests, but I think this change should also be up streamed to transformers main branch. There is still room for improvement.

Yep, I agree this should also be in Transformers. Feel free to open a PR there :slightly_smiling_face:

File: optimum/habana/transformers/trainer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (regisss) [reply optimum/habana/transformers/trainer.py:1655]: The PR looks good to me, do you want to add this to this PR?

[the remark this answers is not in the document]
Comment (yafshar) [reply optimum/habana/transformers/trainer.py:1655]: We can merge this and after I finish my tests resolve the TODO lists. I will follow your opinion.

[the remark this answers is not in the document]
Comment (regisss) [reply optimum/habana/transformers/trainer.py:1655]: Sounds good, merging this one then
